app/com/arpnetworking/metrics/portal/scheduling/impl/JobScheduler.java

```diff
@@ -0,0 +1,225 @@
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.arpnetworking.metrics.portal.scheduling.impl;
+
+import akka.actor.AbstractActorWithTimers;
+import akka.actor.ActorRef;
+import akka.actor.PoisonPill;
+import akka.actor.Props;
+import akka.pattern.PatternsCS;
+import com.arpnetworking.metrics.portal.scheduling.JobRef;
+import com.arpnetworking.steno.Logger;
+import com.arpnetworking.steno.LoggerFactory;
+import models.internal.scheduling.Job;
+import scala.Serializable;
+import scala.concurrent.duration.Duration;
+import scala.concurrent.duration.FiniteDuration;
+
+import java.time.Clock;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * An actor that executes {@link Job}s.
+ *
+ * @param <T> The type of result produced by the {@link Job}s.
+ *
+ * @author Spencer Pearson (spencerpearson at dropbox dot com)
+ */
+public final class JobScheduler<T> extends AbstractActorWithTimers {
+
+    private final JobRef<T> _jobRef;
+    private final Clock _clock;
+
+    /**
+     * Props factory.
+     *
+     * @param <T> The type of result produced by the {@link JobRef}'s job.
+     * @param jobRef The job to intermittently execute.
+     * @return A new props to create this actor.
+     */
+    public static <T> Props props(final JobRef<T> jobRef) {
+        return props(jobRef, Clock.systemUTC());
+    }
+
+    /**
+     * Props factory.
+     *
+     * @param <T> The type of result produced by the {@link JobRef}'s job.
+     * @param jobRef The job to intermittently execute.
+     * @param clock The clock the scheduler will use, when it ticks, to determine whether it's time to run the next job(s) yet.
+     * @return A new props to create this actor.
+     */
+    protected static <T> Props props(final JobRef<T> jobRef, final Clock clock) {
+        return Props.create(JobScheduler.class, () -> new JobScheduler<>(jobRef, clock));
+    }
+
+    private JobScheduler(final JobRef<T> jobRef, final Clock clock) {
+        _jobRef = jobRef;
+        _clock = clock;
+        timers().startPeriodicTimer("TICK", Tick.INSTANCE, TICK_INTERVAL);
+    }
+
+    /**
+     * Computes the time we should schedule an extra tick for, if we'll need an extra tick before the next naturally occurring one.
+     *
+     * @param now The current time.
+     * @param timeToAwaken The time we want to wake up right after.
+     * @return The time until we should wake up, if that's before the next tick; else {@code empty}.
+     */
+    protected static Optional<FiniteDuration> timeUntilExtraTick(final Instant now, final Instant timeToAwaken) {
```

Comment:
Would it be easier if this just returned the time to sleep which is either the delta+slop or the tick? (so no optional result?)